read-cache.c

```diff
@@ -1266,7 +1266,7 @@ static int add_index_entry_with_check(struct index_state *istate, struct cache_e
 	 */
 	if (istate->cache_nr > 0 &&
 		strcmp(ce->name, istate->cache[istate->cache_nr - 1]->name) > 0)
-		pos = -istate->cache_nr - 1;
+		pos = -1 - istate->cache_nr;
```

Comment:
Sneaky.  Would this be better expressed as something like:

    pos = -(signed)(istate->cache_nr + 1);

or define a new st_? function that captures and explains this?
This ( -pos - 1 ) expression is used lots of places, IIRC.

Reply:
I dunno.

TBH I do not really find it sneaky, as I always thought about this operation as "mirroring around `-1`".

What I *would* find sneaky would be to replace `-istate->cache_nr - 1` by `~istate->cache_nr`, which is equivalent, from a bit-wise point of view.

And all these musings leave out the *real* can of worms that you mentioned with your innocuous suggestion to define a new `st_*()` macro. Especially given your remark that it is used in lots of places. I do not even have a good name for it, and I feel we really need one, other than "bitwise complement" or "shifted negative" or whatever: it is a really important paradigm we use for cases when we either return an index for precise matches, and the bitwise complement when we found an insertion point (or something similar) but no precise match.

Note: I do agree that this is a good idea, and that it should be pursued. I do not even want to know whether there is a realistic chance that this thing could overflow, it scares me too much. Although I have to admit that it sounds a bit unrealistic: to overflow, the index would have to overflow into the signed bit, and the only scenario I can come up with is a >2GB byte array on a 32-bit system.

In any case, going this far would definitely exceed the purpose of this PR, in my mind.

So I'll just leave this as-is ;-)